core: stdcm: cache timetable requirements

eckter opened the pull request:
Fix https://github.com/OpenRailAssociation/osrd/issues/12566
Fix https://github.com/OpenRailAssociation/osrd/issues/12567
Fix https://github.com/OpenRailAssociation/osrd/issues/12565
Fix https://github.com/OpenRailAssociation/osrd/issues/12568

Not handled at all and it may be tricky: support for logging heavy payloads with all context

Missing for now:
- ~full benchmark~ -> really speeds up process once cache is set up -> no oom in editoast for long paths anymore in stdcm
- ~fix tests? some seem to be failing~

Erashin commented on core/src/main/java/fr/sncf/osrd/api/TimetableCacheManager.kt:
```diff
@@ -0,0 +1,122 @@
+
+        val rangeSets = mutableMapOf<ZoneId, RangeSet<Double>>()
+        for (rawReq in fetchRawRequirements(timetableId)) {
+            val spacingReq = SpacingRequirement.fromRJS(rawReq, infra)
```
What about `RoutingRequirements`? I know you handle `WorkSchedules` in the last commit, but it seems weird to not cache the routing reqs as well. Especially since I'll send 'em to you, the paylod you'll receive being:
```kotlin
Requirements(
   val trainsRequirements: Map<String, TrainRequirementsRequest>,
   val workSchedules: WorkSchedulesRequest? = null,
```
Except if I just separate the endpoints to have 3 clean spacing/routing/work_schedules endpoints, but that just means you'll have to do different endpoint calls, doesn't change the question.

eckter replied:
We've never considered routing requirements in stdcm. We'd need an incremental spacing resource generator, which would be tedious to write and slow to run.

And most importantly we would almost never detect any additional conflicts. In the current workflow (stops on closed signals and "transit souple"), it would take some *extremely* unlikely corner case to detect a small additional conflict.

Erashin replied:
Only remaining question would be if we used the routing reqs for the conflict detection endpoint, but that'll be a pb for further along then :)

Erashin commented on core/src/main/java/fr/sncf/osrd/api/TimetableCacheManager.kt:
```diff
@@ -0,0 +1,122 @@
+    private fun fetchRawRequirements(timetableId: TimetableId): Sequence<RJSSpacingRequirement> =
+        sequence {
+            // TODO: Double-check the exact editoast API
+            val endpointPath = "timetable/$timetableId/requirements/"
```
`$timetableId/requirements`, but yeah that's what I'm going for :)

~Edit: might go for 3 endpoints: spacing/routing/work_schedules. TBD.~ Nope. Keeping 2 endpoints: the existing one for `work_schedules` and the new one for both spacing and routing requirements.

Erashin commented on core/src/main/kotlin/fr/sncf/osrd/api/stdcm/STDCMEndpoint.kt:
```diff
@@ -381,32 +365,6 @@ private fun parseSimulationScheduleItems(
 }
 
-/** Sanity check, we assert that the result is not conflicting with the scheduled timetable */
-private fun checkForConflicts(
```
This is for performance gains? We could have compared it to the list of timetable cached requirements.

eckter replied:
We don't have the requirement list anymore, we can't just call the other endpoint as an easy sanity check.

That assert hasn't been triggered since we already run conflict detection checks during post-processing anyway. Before that refacto it helped a lot, now it seems less necessary.

Erashin commented on core/src/main/kotlin/fr/sncf/osrd/conflicts/IncrementalConflictDetector.kt:
```diff
@@ -140,3 +119,22 @@ class IncrementalConflictDetector(requirements: List<Requirements>) {
+
+fun generateSpacingZoneUses(
+    requirements: List<Requirements>
+): Map<ZoneId, TreeMap<Double, Range<Double>>> {
```
`ParsedRequirements`

Erashin commented on core/src/main/kotlin/fr/sncf/osrd/conflicts/IncrementalConflictDetector.kt:
```diff
@@ -34,32 +38,7 @@ data class NoConflictResponse(
-            .toMap()
-    }
-
+class IncrementalConflictDetector(private val spacingZoneUses: ParsedRequirements) {
```
I checked, but I just want to make sure, I'm guessing that the `RequirementId` was not being used here, and is only used in `Conflicts` yeah?

eckter replied:
Yep, never used

Erashin reviewed (commented):
Considering that we already have a working work_schedules endpoint, ~it might be better to create 2 additional endpoints respectively for spacing and routing requirements.~ -> nope, it means we'll double the call time since we'll call the cache 2 times instead of once, so there'll only be one endpoint for both spacing and routing requirements.

Erashin commented:
Owning this issue while @eckter is on vacation. Don't forget to add a flag to disable timetable cache to be able to run tests/benchmarks.

Erashin committed:
editoast: stdcm: replace train requirements by timetable_id in endpoint

Signed-off-by: Erashin <alwenn.charpentier@gmail.com>

flomonster committed:
osrdyne: fix worker key encoding to accept dashes

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>

Khoyo committed:
osrdyne: switch from smallvec to smoll_str

Signed-off-by: Younes Khoudli <younes.khoudli@epita.fr>

clarani committed:
front: drop reload infra system

Previously, if the infrastructure failed to load, it would be retried up to four times before displaying an error message.

This system is no longer useful today and must be removed in order to begin refactoring the infrastructure loading system on the backend.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: move infra in the ScenarioContext

Moving the infra and isInfraLoaded in the ScenarioContext allow us to call useInfraStatus only once in the
ScenarioPage.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: remove useInfraStatus call from StdcmWarningBox

Now, useInfraStatus is called only once in the StdcmView.

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

clarani committed:
front: use new polling endpoint

Signed-off-by: Clara Ni <clara.ni@outlook.fr>
Signed-off-by: Erashin <alwenn.charpentier@gmail.com>

flomonster committed:
editoast: replace infra load by worker load endpoint

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>

clarani committed:
front: drop loading infra when creating a scenario

Signed-off-by: Clara Ni <clara.ni@outlook.fr>

Erashin commented on editoast/src/views/worker_load.rs:
```diff
@@ -0,0 +1,189 @@
+    FetchStatusError(#[from] osrdyne_client::Error),
+}
+
+/// Instructs Core to load an infra
```
`Instructs Core to load a worker`? or `load an infra and a potential timetable`?

Khoyo replied:
`Ensure a worker for the given infra (and stdcm timetable) is ready`

`core` doesn't load workers (`osrdyne` does)

Khoyo reviewed (approved):
Core part lgtm

Erashin committed:
editoast: adapt format worker key with timetable

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>
Signed-off-by: Erashin <alwenn.charpentier@gmail.com>

eckter committed:
core: add timetable cache manager

Signed-off-by: Eloi Charpentier <eloi.charpentier.42@gmail.com>
Co-authored-by: Erashin <alwenn.charpentier@gmail.com>

eckter committed:
core: stdcm: use cache manager for timetables

Signed-off-by: Eloi Charpentier <eloi.charpentier.42@gmail.com>
Co-authored-by: Erashin <alwenn.charpentier@gmail.com>

flomonster reviewed (approved):
Editoast part LGTM

SarahBellaha reviewed (approved):
LGTM 💪

kmer2016 reviewed (approved):
Good ! LGTM for the front part !

Maymanaf reviewed (approved):
LGTM for e2e tests

Erashin committed:
tests: remove load infra call in integration tests

Signed-off-by: Erashin <alwenn.charpentier@gmail.com>

shenriotpro reviewed (approved):
LGTM for tests/ (assuming tests pass)

Erashin committed:
editoast: remove stdcm_log endpoint

Signed-off-by: Erashin <alwenn.charpentier@gmail.com>
Co-authored-by: Florian Amsallem <florian.amsallem@gmail.com>

flomonster committed:
front: e2e: fix expected timing of stdcm queries

Since stdcm now takes into account paced trains some timings are
changed.

Signed-off-by: Florian Amsallem <florian.amsallem@gmail.com>

eckter commented on core/src/main/java/fr/sncf/osrd/api/TimetableCacheManager.kt:
```diff
@@ -0,0 +1,174 @@
+
+    /** Load given timetable ID. */
+    fun load(infraId: String, infra: RawInfra, timetableId: TimetableId) {
+        runBlocking { get(infraId, infra, timetableId) }
```
Do we want it to be blocking? Stdcm request will block anyway when synchronizing on the mutex. And if this doesn't block here we can handle other requests in the meantime

Erashin replied:
When we call it in `WorkerCommand`, we want the pod status to be ready when both the infra and the timetable have been loaded, so keeping it likeso is better.

eckter commented on core/src/main/java/fr/sncf/osrd/api/TimetableCacheManager.kt:
```diff
@@ -0,0 +1,174 @@
+ * Caches train spacing requirements for STDCM. The spacing requirements times are relative to
+ * EPOCH.
+ */
+class TimetableCacheManager(
```
We could add opentelemetry spans

eckter commented on core/src/main/java/fr/sncf/osrd/api/TimetableCacheManager.kt:
```diff
@@ -0,0 +1,174 @@
+        val request =
+            buildRequest(endpointPath, "infra_id=$infraId&page=$page&page_size=$PAGE_SIZE")
+        val response = httpClient.newCall(request).execute()
+        if (!response.isSuccessful) throw UnexpectedHttpResponse(response)
```
We should probably retry in some cases

eckter replied:
Oh and if there's any possible exception in the "fetch" functions, we need to wrap the mutex handling in a `try ... finally` block. Otherwise the workers would be left in an unusable state.

edit: wait actually no, the mutex would be left behind but it wouldn't be locked. It's bad but not *that* bad

eckter commented on core/src/main/java/fr/sncf/osrd/api/TimetableCacheManager.kt:
```diff
@@ -0,0 +1,174 @@
+            .adapter(PaginatedRequirements::class.java)
+}
+
+val EPOCH_ZONED: ZonedDateTime = Instant.EPOCH.atZone(java.time.ZoneId.of("UTC"))
```
Seconds since epoch is a large value, and times are stored as doubles. We need to make sure we keep enough precision. I'll run tests. 

Edit: with the current "seconds since epoch", doubles have a precision of 1e-7. For comparaison, in a 24h range, doubles have a precision of 1e-12. It sounds good enough either way.

The only concern I have left on this is debugging difficulties, very large values are harder to read and to compare.

eckter replied:
Actually it's fine, we switch back to "0 as the earliest departure time" in the stdcm endpoint.

eckter reviewed (commented):
Commenting after it's merge now that I'm back from vacations, I'll probably fix those myself or open issues
